attractions/utils/datetime-utils.js

```diff
@@ -2,11 +2,17 @@ const numberRegex = /\d+/g;
 const formatSpecifierRegex = /%[HMSpPdmyY%]/g;
 const daysInWeek = 7;
 
+/**
+ * @template T1
+ * @template T2
+ * @typedef {T1 extends null ? T1 : T2 extends null ? T1 : T2} NullFirst
```

Comment:
Not sure I can work out from this name what `NullFirst` actually is

Reply:
It's to properly type functions with the following guard:
```ts
if (source == null || destination == null) {
  return source;
}
```
that do not return `null` otherwise. The functions are sometimes called with non-nullable arguments, but the return type was still `Date | null`, which caused errors when using the returned value as non-nullable. This type better represents the implementation.

You can read like so: if the first parameter is `null`, then the function returns `null`. Otherwise, if the second parameter is `null`, then we return the first parameter (as non-nullable). Otherwise, we return the \[type of the\] second parameter (also non-nullable).

I agree the name is not very descriptive, but suggestions are welcome.

Reply:
How about `NullLikeArguments` or `NullFromArguments`? This util is making my head spin, I too cannot express it in words

Reply:
How about `PossiblyFilterNull`? This utility type is not really specific to **arguments** only

Reply:
No, I don't really like any of those names, now that  I think about it. And to be honest, even this utility is a very complicated little snippet if you're not in context of what's happening (i.e. remembering everything we've talked about here). So I think that the healthiest solution is just to make the return type `Date | null` and add appropriate guards. Could this mean we'll need a lot of guards? Yeah, but then maybe we should start considering refactoring this code

Reply:
I don't think it's really the healthiest solution. Types are generally supposed to dictate what implementation is correct, or in this case (since types aren't a part of JavaScript) to document the implementation as much as possible. Writing `Date | null` would be an inaccurate representation of what the function actually does. Perhaps the implementation itself is a little twisted (partially due to the lack of null safety in the whole project) and can be improved, but it is what it is at the moment

Reply:
Alright, let's just go with this for now and then in the future we might refactor the code to remove the need for such complicated types altogether